File: src/datalogger.rs
use crate::current_monitor::{CurrentArray, CurrentType};
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::ops::Deref;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct DataLogger<'a, const N: usize> {
    start_time: Time,
    day_time: u64,
    time_between_logs: u64,
    folder: &'a Path,
    filepath: PathBuf,
    last_logged: Option<u64>,
    day: u64,
    current_types: &'a [CurrentType; N],
    names: &'a [&'a str; N],
}

impl<'a, const N: usize> DataLogger<'a, N> {
    fn create_new_file(
        time: Time,
        folder: &Path,
        current_types: &[CurrentType; N],
        names: &[&str; N],
    ) -> PathBuf {
        if !folder.exists() {
            std::fs::create_dir(folder).unwrap();
        }
        let day = time.get_day();
        let main_filename = format!("D{day}");
        let mut i = 1;
        let path = loop {
            let test_filename = if i == 1 {
                format!("{main_filename}.csv")
            } else {
                format!("{main_filename}_{i}.csv")
            };
            let path = folder.join(&test_filename);
            if !path.exists() {
                break path;
            }
            i += 1;
        };
        let mut file = File::create(&path).expect("Error encountered while creating file!");
        let mut first_line = "Time".to_string();
        for (name, current_type) in names.iter().zip(current_types) {
            first_line.push_str(&format!(",{name} ({current_type})"))
        }
        writeln!(file, "{}", first_line).unwrap();
        path
    }

    pub fn new(
        time_between_logs: u64,
        folder: &'a Path,
        current_types: &'a [CurrentType; N],
        names: &'a [&'a str; N],
    ) -> DataLogger<'a, N> {
        let start_time = Time::new_unix_time();
        DataLogger {
            start_time,
            day_time: start_time.0 % 86400,
            time_between_logs,
            folder,
            filepath: DataLogger::create_new_file(start_time, folder, current_types, names),
            last_logged: None,
            day: start_time.get_day(),
            current_types,
            names,
        }
    }

    pub fn tick(&mut self, current_array: &CurrentArray<N>) {
        let unix = Time::new_unix_time();
        if let Some(t) = self.last_logged {
            if self.start_time < unix && self.day != unix.get_day() {
                self.filepath =
                    DataLogger::create_new_file(unix, self.folder, self.current_types, self.names);
                self.day = unix.get_day();
            }
            if *unix < t + self.time_between_logs {
                return;
            }
        }
        let time = chrono::Utc::now().time();
        let mut file = OpenOptions::new()
            .append(true)
            .open(&self.filepath)
            .unwrap();
        writeln!(
            file,
            "{},{}",
            time.format("T%H%M%SZ"),
            current_array.generate_line()
        )
        .unwrap();
        self.last_logged = Some(*unix);
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct Time(u64);

impl Time {
    fn new_unix_time() -> Time {
        Time(
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        )
    }

    fn get_day(&self) -> u64 {
        (self.0 as f64 / 86400.).floor() as u64
    }
}

impl Deref for Time {
    type Target = u64;
    fn deref(&self) -> &u64 {
        &self.0
    }
}

File: python/datalogger.py
"""Handles datalogging."""
from config import Config
from datetime import datetime
from pathlib import Path
import time


class DataLogger:
    """Logs the data."""

    def __init__(self, config: Config, freq: int, folder: Path):
        """Create the file to log in and fills in the titles."""
        self.logger = config.logger
        folder = config.path / folder
        if not folder.is_dir():
            folder.mkdir()
        self.freq = freq
        self.folder = folder
        self.names = config.names
        self.current_types = config.current_types
        self.start_time = time.time()
        self._new_file()
        self.last_tick = None

    def _new_file(self):
        header = 'Time' + ''.join([f',{n}({t.name})' for (n, t) in
                                   zip(self.names, self.current_types)])
        self.day = int(time.time() // 86400)
        root_filename = f'D{self.day}'
        i = 1
        while True:
            if i == 1:
                test_filename = f'{root_filename}.csv'
            else:
                test_filename = f'{root_filename}_{i}.csv'
            path = self.folder / test_filename
            file_header = None
            if path.is_file():
                with open(path, 'r') as fp:
                    file_header = fp.readline().replace('\n', '')
                if file_header == header:
                    self.fp = path
                    break
            if not path.is_file():
                self.fp = path
                with open(path, 'x') as fp:
                    fp.write(header)
                break
            i += 1

    def tick(self, currents: [float]):
        """Log the data if enough time has passed."""
        this_tick = time.time() // self.freq
        if self.last_tick is None or self.last_tick < this_tick:
            self.last_tick = this_tick
            if int(time.time() // 86400) != self.day:
                self._new_file()
            self._log_to_file(currents)

    def _log_to_file(self, currents: [float]):
        with open(self.fp, 'a') as fp:
            mes = str(datetime.now().replace(microsecond=0).isoformat())
            mes += ''.join([f',{c}' for c in currents])
            fp.write(f'\n{mes}')

    def add_metadata(self, metadata: str):
        """Add a bit of metadata to the datalog."""
        with open(self.fp, 'a') as fp:
            fp.write(f',{metadata}')
